README
NAME
    GraphViz2::Abstract::Node - Deal with nodes independent of a Graph

VERSION
    version 0.001002

SYNOPSIS
        use GraphViz2::Abstract::Node;

        my $node = GraphViz2::Abstract::Node->new(
                color =>  ... ,
                id    =>  ... ,
                label =>  ... ,
        );

        # Mutate $node

        $node->label("Asdft");

        my $fillcolor = $node->fillcolor(); # Knows that the fill color is light grey despite never setting it.

        # Later:

        $graph->add_node(%{ $node->as_hash }); # Adds only the data that is not the same as GraphViz's defaults
        $graph->add_node(%{ $node->as_canon_hash }); # Adds all the data, including hardcoded defaults

DESCRIPTION
    Working with GraphViz2, I found myself frequently needing shared styles
    for things, and I often had trouble knowing which fields were and
    weren't valid for given things, for instance: "Nodes".

    Its reasonably straight forward to ask the question "What is the
    attribute "foo" applicable to" using the GraphViz website, but much
    harder to know "What are all the attributes applicable to "foo"".

    Let alone work with them in a user friendly way from code.

  Naming Rationale
    I tried to choose a name that was not so likely to threaten GraphViz2 if
    GraphViz2 wanted to make a different variation of what I'm doing, but as
    part of GraphViz2 itself.

    As such, I plan on a few "::Abstract" things, that aim to be stepping
    stones for dealing with complex data independent of "GraphViz2", but in
    such a way that they make importing that data into "GraphViz2" easy.

METHODS
  "as_hash"
    This method returns all the values of all properties that DIFFER from
    the defaults.

    e.g.

        Node->new( color => 'black' )->as_hash();

    Will return an empty list, as the default color is normally black.

    See also how special constants work in

  "as_canon_hash"
    This method returns all the values of all properties, INCLUDING
    defaults.

    e.g.

        Node->new( color => 'black' )->as_canon_hash();

    Will return a very large list containing all the properties that we know
    the default values for.

    See also how special constants work in

ATTRIBUTES
  "URL"
    Default: "none"

  "area"
    Default: 1.0

  "color"
    Default: "black"

  "colorscheme"
    Default: ""

  "comment"
    Default: ""

  "distortion"
    Default: 0.0

  "fillcolor"
    Default: "lightgrey"

  "fixedsize"
    Default: "false"

  "fontcolor"
    Default: "black"

  "fontname"
    Default: "Times-Roman"

  "fontsize"
    Default: 14.0

  "gradientangle"
    Default: ""

  "group"
    Default: ""

  "height"
    Default: 0.5

  "href"
    Default: ""

  "id"
    Default: ""

  "image"
    Default: ""

  "imagescale"
    Default: "false" ( Yes, really! )

  "label"
    Default: "\\N" ( Appears to be a magic value for GraphViz )

  "labelloc"
    Default: "c"

  "layer"
    Default: ""

  "margin"
    Default: "unknown" ( Due to being render device specific defaults )

  "nojustify"
    Default: "false"

  "ordering"
    Default: ""

  "orientation"
    Default: 0.0

  "penwidth"
    Default: 1.0

  "peripheries"
    Default: "unknown"

  "pos"
    Default: "unknown"

  "rects"
    Default: "unknown"

  "regular"
    Default: "false"

  "root"
    Default: "false"

  "samplepoints"
    Default: "unknown"

    Reason: Dependent on render device.

  "shape"
    Default: "ellipse"

  "shapefile"
    Default: ""

  "showboxes"
    Default: 0

  "sides"
    Default: 4

  "skew"
    Default: 0.0

  "sortv"
    Default: 0

  "style"
    Default: ""

  "target"
    Default: "none"

  "tooltip"
    Default: ""

  "vertices"
    Default: "unknown"

  "width"
    Default: 0.75

  "xlabel"
    Default: ""

  "xlp"
    Default: ""

  "z"
    Default: 0.0

AUTHOR
    Kent Fredric <kentfredric@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for GraphViz2-Abstract-Node

{{$NEXT}}

0.002000 2013-12-01T05:47:00Z
 - Factor out constants into GV2:A:Util::Constants

 [Dependencies::Stats]
 - Dependencies changed, see Changes.deps{,.all,.dev} for details
 - develop: +2 ↑1 -2
 - runtime: +1 -1

0.001001 2013-11-30T09:18:32Z
 - Add URL and showboxes fields.

 [Dependencies::Stats]
 - Dependencies changed, see Changes.deps{,.all,.dev} for details
 - develop: ↑1 -1

0.001000 2013-11-29T17:59:27Z
 - First version.



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4203"
  },
  "configure_requires" => {
    "Module::Build" => "0.4203"
  },
  "dist_abstract" => "Deal with nodes independent of a Graph",
  "dist_author" => [
    "Kent Fredric <kentfredric\@gmail.com>"
  ],
  "dist_name" => "GraphViz2-Abstract-Node",
  "dist_version" => "0.001002",
  "license" => "perl",
  "module_name" => "GraphViz2::Abstract::Node",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Class::Tiny" => 0,
    "GraphViz2::Abstract::Util::Constants" => 0,
    "Scalar::Util" => 0,
    "perl" => "5.014002",
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "GraphViz2" => 0,
    "Path::Tiny" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "1.001002",
    "parent" => 0
  }
);


my %fallback_build_requires = (
  "GraphViz2" => 0,
  "Module::Build" => "0.4203",
  "Path::Tiny" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "1.001002",
  "parent" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


maint-travis-ci/install_deps_early.pl
#!/usr/bin/env perl
use strict;
use warnings;
use utf8;

use FindBin;
use lib "$FindBin::Bin/lib";
use tools;

if ( not env_exists('TRAVIS') ) {
  diag('Is not running under travis!');
  exit 1;
}
if ( not env_exists('STERILIZE_ENV') ) {
  diag("\e[31STERILIZE_ENV is not set, skipping, because this is probably Travis's Default ( and unwanted ) target");
  exit 0;
}
my (@params) = qw[ --quiet --notest --mirror http://cpan.metacpan.org/ --no-man-pages ];
my ($branch) = $ENV{TRAVIS_BRANCH};
my ($prefix) = './.travis_early_installdeps.';
$branch =~ s{/}{_}g;
my ($depsfile)   = ( $prefix . $branch );
my ($paramsfile) = ( $prefix . 'params.' . $branch );

if ( not( -e $depsfile and -f $depsfile ) ) {
  diag("\e[31m$depsfile does not exist, no extra deps\e[0m");
  exit 0;
}

my (@deps) = split /\n/, do {
  open my $fh, '<', $depsfile;
  local $/ = undef;
  scalar <$fh>;
};
if ( -e $paramsfile and -f $paramsfile ) {
  push @params, split /\n/, do {
    open my $fh, '<', $paramsfile;
    local $/ = undef;
    scalar <$fh>;
  };
}
cpanm( @params, @deps );
exit 0;


lib/GraphViz2/Abstract/Node.pm
use strict;
use warnings;
use utf8;

package GraphViz2::Abstract::Node;

# ABSTRACT: Deal with nodes independent of a Graph

use GraphViz2::Abstract::Util::Constants;

our @CARP_NOT;

=head1 SYNOPSIS

    use GraphViz2::Abstract::Node;

    my $node = GraphViz2::Abstract::Node->new(
            color =>  ... ,
            id    =>  ... ,
            label =>  ... ,
    );

    # Mutate $node

    $node->label("Asdft");

    my $fillcolor = $node->fillcolor(); # Knows that the fill color is light grey despite never setting it.

    # Later:

    $graph->add_node(%{ $node->as_hash }); # Adds only the data that is not the same as GraphViz's defaults
    $graph->add_node(%{ $node->as_canon_hash }); # Adds all the data, including hardcoded defaults

=head1 DESCRIPTION

Working with GraphViz2, I found myself frequently needing shared styles for things, and I often had trouble knowing
which fields were and weren't valid for given things, for instance: C<Nodes>.

Its reasonably straight forward to ask the question "What is the attribute C<foo> applicable to" using the GraphViz website,
but much harder to know "What are all the attributes applicable to C<foo>".

Let alone work with them in a user friendly way from code.

=head2 Naming Rationale

I tried to choose a name that was not so likely to threaten GraphViz2 if GraphViz2 wanted to make a different
variation of what I'm doing, but as part of GraphViz2 itself.

As such, I plan on a few C<::Abstract> things, that aim to be stepping stones for dealing with complex data independent of C<GraphViz2>,
but in such a way that they make importing that data into C<GraphViz2> easy.

=cut

use Class::Tiny {
  URL           => NONE,
  area          => 1.0,
  color         => 'black',
  colorscheme   => EMPTY_STRING,
  comment       => EMPTY_STRING,
  distortion    => 0.0,
  fillcolor     => 'lightgrey',
  fixedsize     => FALSE,
  fontcolor     => 'black',
  fontname      => 'Times-Roman',
  fontsize      => 14.0,
  gradientangle => EMPTY_STRING,
  group         => EMPTY_STRING,
  height        => 0.5,
  href          => EMPTY_STRING,
  id            => EMPTY_STRING,
  image         => EMPTY_STRING,
  imagescale    => FALSE,
  label         => q[\\N],
  labelloc      => q[c],
  layer         => EMPTY_STRING,
  margin        => UNKNOWN,
  nojustify     => FALSE,
  ordering      => EMPTY_STRING,
  orientation   => 0.0,
  penwidth      => 1.0,
  peripheries   => UNKNOWN,
  pos           => UNKNOWN,
  rects         => UNKNOWN,
  regular       => FALSE,
  root          => FALSE,
  samplepoints  => UNKNOWN,
  shape         => 'ellipse',
  shapefile     => EMPTY_STRING,
  showboxes     => 0,
  sides         => 4,
  skew          => 0.0,
  sortv         => 0,
  style         => EMPTY_STRING,
  target        => NONE,
  tooltip       => EMPTY_STRING,
  vertices      => UNKNOWN,
  width         => 0.75,
  xlabel        => EMPTY_STRING,
  xlp           => EMPTY_STRING,
  z             => 0.0,
};

=attr C<URL>

Default: L<< C<none>|GraphViz2::Abstract::Util::Constants/NONE >>

=attr C<area>

Default: C<1.0>

=attr C<color>

Default: C<"black">

=attr C<colorscheme>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<comment>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<distortion>

Default: C<0.0>

=attr C<fillcolor>

Default: C<"lightgrey">

=attr C<fixedsize>

Default: L<< C<false>|GraphViz2::Abstract::Util::Constants/FALSE >>

=attr C<fontcolor>

Default: C<"black">

=attr C<fontname>

Default: C<"Times-Roman">

=attr C<fontsize>

Default: C<14.0>

=attr C<gradientangle>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<group>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<height>

Default: C<0.5>

=attr C<href>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<id>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<image>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<imagescale>

Default: L<< C<false>|GraphViz2::Abstract::Util::Constants/FALSE >>  ( Yes, really! )

=attr C<label>

Default: C<"\\N"> ( Appears to be a magic value for GraphViz )

=attr C<labelloc>

Default: C<"c">

=attr C<layer>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<margin>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>  ( Due to being render device specific defaults )

=attr C<nojustify>

Default: L<< C<false>|GraphViz2::Abstract::Util::Constants/FALSE >>

=attr C<ordering>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<orientation>

Default: C<0.0>

=attr C<penwidth>

Default: C<1.0>

=attr C<peripheries>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>

=attr C<pos>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>

=attr C<rects>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>

=attr C<regular>

Default: L<< C<false>|GraphViz2::Abstract::Util::Constants/FALSE >>

=attr C<root>

Default: L<< C<false>|GraphViz2::Abstract::Util::Constants/FALSE >>

=attr C<samplepoints>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>

Reason: Dependent on render device.

=attr C<shape>

Default: C<"ellipse">

=attr C<shapefile>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<showboxes>

Default: C<0>

=attr C<sides>

Default: C<4>

=attr C<skew>

Default: C<0.0>

=attr C<sortv>

Default: C<0>

=attr C<style>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<target>

Default: L<< C<none>|GraphViz2::Abstract::Util::Constants/NONE >>

=attr C<tooltip>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<vertices>

Default: L<< C<unknown>|GraphViz2::Abstract::Util::Constants/UNKNOWN >>

=attr C<width>

Default: C<0.75>

=attr C<xlabel>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<xlp>

Default: L<< C<"">|GraphViz2::Abstract::Util::Constants/EMPTY_STRING >>

=attr C<z>

Default: C<0.0>

=cut

use Scalar::Util qw(blessed);
use Scalar::Util qw(refaddr);

sub _is_equal {
  my ( $self, $a_ref, $b_ref ) = @_;

  return   if defined $a_ref     and not defined $b_ref;
  return   if not defined $a_ref and defined $b_ref;
  return 1 if not defined $a_ref and not defined $b_ref;

  ## A and B are both defined.

  return if not ref $a_ref and ref $b_ref;
  return if ref $a_ref and not $b_ref;

  if ( not ref $a_ref and not ref $b_ref ) {
    return $a_ref eq $b_ref;
  }

  ##  A and B are both refs.
  return refaddr $a_ref eq refaddr $b_ref;
}

sub _is_magic {
  my ( $self, $value ) = @_;
  return if not defined $value;
  return if not ref $value;
  my $addr = refaddr $value;
  return 1 if $addr eq refaddr UNKNOWN;
  return 1 if $addr eq refaddr NONE;
  return;
}

sub _foreach_attr {
  my ( $self, $callback ) = @_;
  if ( not blessed($self) ) {
    require Carp;
    local @CARP_NOT = 'GraphViz2::Abstract::Node';
    Carp::croak('Can\'t call as_hash on a class');
  }
  my $class    = blessed($self);
  my @attrs    = Class::Tiny->get_all_attributes_for($class);
  my $defaults = Class::Tiny->get_all_attribute_defaults_for(__PACKAGE__);
  for my $attr (@attrs) {
    my $value       = $self->$attr();
    my $has_default = exists $defaults->{$attr};
    my $default;
    if ($has_default) {
      $default = $defaults->{$attr};
    }
    $callback->( $attr, $value, $has_default, $default );
  }
  return $self;
}

=method C<as_hash>

This method returns all the values of all properties that B<DIFFER> from the defaults.

e.g.

    Node->new( color => 'black' )->as_hash();

Will return an empty list, as the default color is normally black.

See also L<< how special constants work in|GraphViz2::Abstract::Util::Constants/CONSTANTS >>

=cut

sub as_hash {
  my ($self) = @_;
  my %output;

  $self->_foreach_attr(
    sub {
      my ( $attr, $value, $has_default, $default ) = @_;
      if ( not $has_default ) {
        return if $self->_is_magic($value);
        $output{$attr} = $value;
        return;
      }
      return if $self->_is_equal( $value, $default );
      return if $self->_is_magic($value);
      $output{$attr} = $value;
    }
  );
  return \%output;
}

=method C<as_canon_hash>

This method returns all the values of all properties, B<INCLUDING> defaults.

e.g.

    Node->new( color => 'black' )->as_canon_hash();

Will return a very large list containing all the properties that we know the default values for.

See also L<< how special constants work in|GraphViz2::Abstract::Util::Constants/CONSTANTS >>

=cut

sub as_canon_hash {
  my ($self) = @_;
  my %output;
  $self->_foreach_attr(
    sub {
      my ( $attr, $value, $has_default, $default ) = @_;
      return if $self->_is_magic($value);
      $output{$attr} = $value;
    }
  );
  return \%output;

}


1;
